{"content":"<div><div class=\"content\">\n<p>Exceptions are used to indicate that an error has occurred while running the program. Exception objects that describe an error are created and then <em>thrown</em> with the <a href=\"../../language-reference/keywords/throw\" data-linktype=\"relative-path\">throw</a> keyword. The runtime then searches for the most compatible exception handler.  </p>\n<p> Programmers should throw exceptions when one or more of the following conditions are true:  </p>\n<ul>\n<li><p>The method cannot complete its defined functionality.  </p>\n<p> For example, if a parameter to a method has an invalid value:  </p>\n<pre><code class=\"lang-cs\" name=\"csProgGuideExceptions#12\">static void CopyObject(SampleClass original)\n{\n    if (original == null)\n    {\n        throw new System.ArgumentException(&quot;Parameter cannot be null&quot;, &quot;original&quot;);\n    }\n\n}\n</code></pre></li>\n<li><p>An inappropriate call to an object is made, based on the object state.  </p>\n<p> One example might be trying to write to a read-only file. In cases where an object state does not allow an operation, throw an instance of <a class=\"xref\" href=\"../../../../api/system.invalidoperationexception\" data-linktype=\"relative-path\">InvalidOperationException</a> or an object based on a derivation of this class. This is an example of a method that throws an <a class=\"xref\" href=\"../../../../api/system.invalidoperationexception\" data-linktype=\"relative-path\">InvalidOperationException</a> object:  </p>\n<pre><code class=\"lang-cs\" name=\"csProgGuideExceptions#13\">class ProgramLog\n{\n    System.IO.FileStream logFile = null;\n    void OpenLog(System.IO.FileInfo fileName, System.IO.FileMode mode) {}\n\n    void WriteLog()\n    {\n        if (!this.logFile.CanWrite)\n        {\n            throw new System.InvalidOperationException(&quot;Logfile cannot be read-only&quot;);\n        }\n        // Else write data to the log and return.\n    }\n}\n</code></pre></li>\n<li><p>When an argument to a method causes an exception.  </p>\n<p> In this case, the original exception should be caught and an <a class=\"xref\" href=\"../../../../api/system.argumentexception\" data-linktype=\"relative-path\">ArgumentException</a> instance should be created. The original exception should be passed to the constructor of the <a class=\"xref\" href=\"../../../../api/system.argumentexception\" data-linktype=\"relative-path\">ArgumentException</a> as the <a class=\"xref\" href=\"../../../../api/system.exception#System_Exception_InnerException_\" data-linktype=\"relative-path\">InnerException</a> parameter:  </p>\n<pre><code class=\"lang-cs\" name=\"csProgGuideExceptions#14\">static int GetValueFromArray(int[] array, int index)\n{\n    try\n    {\n        return array[index];\n    }\n    catch (System.IndexOutOfRangeException ex)\n    {\n        System.ArgumentException argEx = new System.ArgumentException(&quot;Index is out of range&quot;, &quot;index&quot;, ex);\n        throw argEx;\n    }\n}\n</code></pre></li>\n</ul>\n<p> Exceptions contain a property named <a class=\"xref\" href=\"../../../../api/system.exception#System_Exception_StackTrace_\" data-linktype=\"relative-path\">StackTrace</a>. This string contains the name of the methods on the current call stack, together with the file name and line number where the exception was thrown for each method. A <a class=\"xref\" href=\"../../../../api/system.exception#System_Exception_StackTrace_\" data-linktype=\"relative-path\">StackTrace</a> object is created automatically by the common language runtime (CLR) from the point of the <code>throw</code> statement, so that exceptions must be thrown from the point where the stack trace should begin.  </p>\n<p> All exceptions contain a property named <a class=\"xref\" href=\"../../../../api/system.exception#System_Exception_Message_\" data-linktype=\"relative-path\">Message</a>. This string should be set to explain the reason for the exception. Note that information that is sensitive to security should not be put in the message text. In addition to <a class=\"xref\" href=\"../../../../api/system.exception#System_Exception_Message_\" data-linktype=\"relative-path\">Message</a>, <a class=\"xref\" href=\"../../../../api/system.argumentexception\" data-linktype=\"relative-path\">ArgumentException</a> contains a property named <a class=\"xref\" href=\"../../../../api/system.argumentexception#System_ArgumentException_ParamName_\" data-linktype=\"relative-path\">ParamName</a> that should be set to the name of the argument that caused the exception to be thrown. In the case of a property setter, <a class=\"xref\" href=\"../../../../api/system.argumentexception#System_ArgumentException_ParamName_\" data-linktype=\"relative-path\">ParamName</a> should be set to <code>value</code>.  </p>\n<p> Public and protected methods members should throw exceptions whenever they cannot complete their intended functions. The exception class that is thrown should be the most specific exception available that fits the error conditions. These exceptions should be documented as part of the class functionality, and derived classes or updates to the original class should retain the same behavior for backward compatibility.  </p>\n<h2 id=\"things-to-avoid-when-throwing-exceptions\">Things to Avoid When Throwing Exceptions</h2>\n<p> The following list identifies practices to avoid when throwing exceptions:  </p>\n<ul>\n<li><p>Exceptions should not be used to change the flow of a program as part of ordinary execution. Exceptions should only be used to report and handle error conditions.  </p>\n</li>\n<li><p>Exceptions should not be returned as a return value or parameter instead of being thrown.  </p>\n</li>\n<li><p>Do not throw <a class=\"xref\" href=\"../../../../api/system.exception\" data-linktype=\"relative-path\">System.Exception</a>, <a class=\"xref\" href=\"../../../../api/system.systemexception\" data-linktype=\"relative-path\">System.SystemException</a>, <a class=\"xref\" href=\"../../../../api/system.nullreferenceexception\" data-linktype=\"relative-path\">System.NullReferenceException</a>, or <a class=\"xref\" href=\"../../../../api/system.indexoutofrangeexception\" data-linktype=\"relative-path\">System.IndexOutOfRangeException</a> intentionally from your own source code.  </p>\n</li>\n<li><p>Do not create exceptions that can be thrown in debug mode but not release mode. To identify run-time errors during the development phase, use Debug Assert instead.  </p>\n</li>\n</ul>\n<h2 id=\"defining-exception-classes\">Defining Exception Classes</h2>\n<p> Programs can throw a predefined exception class in the <a class=\"xref\" href=\"../../../../api/system\" data-linktype=\"relative-path\">System</a> namespace (except where previously noted), or create their own exception classes by deriving from <a class=\"xref\" href=\"../../../../api/system.exception\" data-linktype=\"relative-path\">Exception</a>. The derived classes should define at least four constructors: one default constructor, one that sets the message property, and one that sets both the <a class=\"xref\" href=\"../../../../api/system.exception#System_Exception_Message_\" data-linktype=\"relative-path\">Message</a> and <a class=\"xref\" href=\"../../../../api/system.exception#System_Exception_InnerException_\" data-linktype=\"relative-path\">InnerException</a> properties. The fourth constructor is used to serialize the exception. New exception classes should be serializable. For example:  </p>\n<pre><code class=\"lang-cs\" name=\"csProgGuideExceptions#15\">[Serializable()]\npublic class InvalidDepartmentException : System.Exception\n{\n    public InvalidDepartmentException() : base() { }\n    public InvalidDepartmentException(string message) : base(message) { }\n    public InvalidDepartmentException(string message, System.Exception inner) : base(message, inner) { }\n\n    // A constructor is needed for serialization when an\n    // exception propagates from a remoting server to the client. \n    protected InvalidDepartmentException(System.Runtime.Serialization.SerializationInfo info,\n        System.Runtime.Serialization.StreamingContext context) { }\n}\n</code></pre><p> New properties should only be added to the exception class when the data they provide is useful to resolving the exception. If new properties are added to the derived exception class, <code>ToString()</code> should be overridden to return the added information.  </p>\n<h2 id=\"c-language-specification\">C# Language Specification</h2>\n<p>For more information, see the <a href=\"../../language-reference/language-specification\" data-linktype=\"relative-path\">C# Language Specification</a>. The language specification is the definitive source for C# syntax and usage.</p>\n<h2 id=\"see-also\">See Also</h2>\n<p> <a href=\"../index\" data-linktype=\"relative-path\">C# Programming Guide</a><br> <a href=\"index\" data-linktype=\"relative-path\">Exceptions and Exception Handling</a><br> <a href=\"http://msdn.microsoft.com/library/f7d68675-be06-40fb-a555-05f0c5a6f66b\" data-linktype=\"external\">Exception Hierarchy</a><br> <a href=\"exception-handling\" data-linktype=\"relative-path\">Exception Handling</a></p>\n</div></div>","outputRootRelativePath":"../../../../","pageMetadata":"<meta name=\"author\" content=\"BillWagner\">\r\n<meta name=\"ms.author\" content=\"wiwagn\">\r\n<meta name=\"manager\" content=\"wpickett\">\r\n<meta name=\"breadcrumb_path\" content=\"../toc2.json\">\r\n<meta name=\"ms.assetid\" content=\"6bbba495-a115-4c6d-90cc-1f4d7b5f39e2\">\r\n<meta name=\"translation.priority.ht\" content=\"cs-cz\">\r\n<meta name=\"translation.priority.ht\" content=\"de-de\">\r\n<meta name=\"translation.priority.ht\" content=\"es-es\">\r\n<meta name=\"translation.priority.ht\" content=\"fr-fr\">\r\n<meta name=\"translation.priority.ht\" content=\"it-it\">\r\n<meta name=\"translation.priority.ht\" content=\"ja-jp\">\r\n<meta name=\"translation.priority.ht\" content=\"ko-kr\">\r\n<meta name=\"translation.priority.ht\" content=\"pl-pl\">\r\n<meta name=\"translation.priority.ht\" content=\"pt-br\">\r\n<meta name=\"translation.priority.ht\" content=\"ru-ru\">\r\n<meta name=\"translation.priority.ht\" content=\"tr-tr\">\r\n<meta name=\"translation.priority.ht\" content=\"zh-cn\">\r\n<meta name=\"translation.priority.ht\" content=\"zh-tw\">\r\n<meta name=\"caps.latest.revision\" content=\"28\">\r\n<meta name=\"ms.topic\" content=\"article\">\r\n<meta name=\"ms.date\" content=\"2015-07-20\">\r\n<meta name=\"ms.technology\" content=\"devlang-csharp\">\r\n<meta name=\"ms.prod\" content=\".net\">\r\n<meta name=\"search.ms_sitename\" content=\"Docs\">\r\n<meta name=\"search.ms_docsetname\" content=\"docs-internal\">\r\n<meta name=\"version\" content=\"0\">\r\n<meta name=\"locale\" content=\"en-us\">\r\n<meta name=\"site_name\" content=\"Docs\">\r\n<meta name=\"search.ms_product\" content=\"MSDN\">\r\n<meta name=\"depot_name\" content=\"MSDN.docs-internal\">\r\n<meta name=\"updated_at\" content=\"2017-03-15 06:14 PM\">\r\n<meta name=\"gitcommit\" content=\"https://github.com/dotnet/docs-internal/blob/3115359729431ecbcc16058612da59b85621fa0f/docs/csharp/programming-guide/exceptions/creating-and-throwing-exceptions.md\">\r\n<meta name=\"original_content_git_url\" content=\"https://github.com/dotnet/docs-internal/blob/master/docs/csharp/programming-guide/exceptions/creating-and-throwing-exceptions.md\">\r\n<meta name=\"document_id\" content=\"03118812-f5ce-5295-c0ea-ee5c71d32e3d\">\r\n<meta name=\"pagetype\" content=\"Conceptual\">\r\n<meta name=\"toc_rel\" content=\"toc2.json\">\r\n","rawMetadata":{"author":"BillWagner","ms.author":"wiwagn","manager":"wpickett","breadcrumb_path":"/dotnet/toc.json","title":"Creating and Throwing Exceptions (C# Programming Guide) | Microsoft Docs","ms.assetid":"6bbba495-a115-4c6d-90cc-1f4d7b5f39e2","translation.priority.ht":["cs-cz","de-de","es-es","fr-fr","it-it","ja-jp","ko-kr","pl-pl","pt-br","ru-ru","tr-tr","zh-cn","zh-tw"],"caps.latest.revision":28,"ms.topic":"article","dev_langs":["csharp"],"helpviewer_keywords":["catching exceptions [C#]","throwing exceptions [C#]","exceptions [C#], creating","exceptions [C#], throwing"],"ms.date":"2015-07-20","ms.technology":["devlang-csharp"],"ms.prod":".net","search.ms_sitename":"Docs","search.ms_docsetname":"docs-internal","version":0,"_op_canonicalUrlPrefix":"https://docs.microsoft.com/en-us/dotnet-internal/","locale":"en-us","site_name":"Docs","search.ms_product":"MSDN","_op_openToPublicContributors":false,"depot_name":"MSDN.docs-internal","_op_gitCommitHistory":[{"author_name":"Bill Wagner","author_email":"wiwagn@microsoft.com","committer_name":"GitHub","comitter_email":"noreply@github.com","message":"Reorganize the C# Guide TOC (#1318)","commit_sha":"3115359729431ecbcc16058612da59b85621fa0f","commit_date":"2017-01-23 23:20:14 -0500"},{"author_name":"Maira Wenzel","author_email":"mairaw@microsoft.com","committer_name":"GitHub","comitter_email":"noreply@github.com","message":"Metadata updates (#1313)","commit_sha":"90fe68f7f3c4b46502b5d3770b1a2d57c6af748a","commit_date":"2016-12-09 12:40:41 -0800"},{"author_name":"Maira Wenzel","author_email":"mairaw@microsoft.com","committer_name":"Bill Wagner","comitter_email":"wiwagn@microsoft.com","message":":books: migration changes (#1235)","commit_sha":"b828bb1d6c8fb750ad9ef34f8a7a1b7d2574f4c6","commit_date":"2016-11-15 15:08:11 -0500"}],"updated_at":"2017-03-15 06:14 PM","_op_gitContributorInformation":{"author":{"profile_url":"https://github.com/BillWagner","display_name":"Bill Wagner","id":"493969"},"contributors":[{"profile_url":"https://github.com/BillWagner","display_name":"Bill Wagner"},{"profile_url":"https://github.com/mairaw","display_name":"Maira Wenzel","id":"12971179"}],"update_at":"3/15/2017"},"gitcommit":"https://github.com/dotnet/docs-internal/blob/3115359729431ecbcc16058612da59b85621fa0f/docs/csharp/programming-guide/exceptions/creating-and-throwing-exceptions.md","original_content_git_url":"https://github.com/dotnet/docs-internal/blob/master/docs/csharp/programming-guide/exceptions/creating-and-throwing-exceptions.md","open_to_public_contributors":false,"content_git_url":"https://github.com/dotnet/docs/blob/master/docs/csharp/programming-guide/exceptions/creating-and-throwing-exceptions.md","document_id":"03118812-f5ce-5295-c0ea-ee5c71d32e3d","layout":"Conceptual","pagetype":"Conceptual","canonical_url":"https://docs.microsoft.com/en-us/dotnet/articles/csharp/programming-guide/exceptions/creating-and-throwing-exceptions","toc_asset_id":"articles/csharp/programming-guide/exceptions/toc.json","toc_rel":"toc.json","_op_ogTitle":"Creating and Throwing Exceptions (C# Programming Guide)","_op_displayDate":null,"_op_displayDate_source":null,"_op_wordCount":749,"_op_rawTitle":"<h1 id=\"creating-and-throwing-exceptions-c-programming-guide\" sourcefile=\"docs/csharp/programming-guide/exceptions/creating-and-throwing-exceptions.md\" sourcestartlinenumber=\"39\" sourceendlinenumber=\"39\">Creating and Throwing Exceptions (C# Programming Guide)</h1>","_op_searchScope":[".NET"],"_op_canonicalUrl":"https://docs.microsoft.com/en-us/dotnet/articles/csharp/programming-guide/exceptions/creating-and-throwing-exceptions","fileRelativePath":"articles/csharp/programming-guide/exceptions/creating-and-throwing-exceptions.html"},"themesRelativePathToOutputRoot":"_themes/"}